scripts/src/blockchain/channelsPool.ts

```diff
@@ -0,0 +1,59 @@
+import {KeyPair} from "./keyPair";
+import {ChannelBusyError} from "../errors";
+
+export class ChannelsPool {
+
+	private readonly channels: Channel[] = [];
+
+	constructor(seeds: string[]) {
+		this.channels = seeds.map(seed => ({keyPair: KeyPair.fromSeed(seed), state: 'free' as ChannelState}));
+	}
+
+	private getFreeChannels(): Channel[] {
+		return this.channels.filter(channel => channel.state == 'free');
+	}
+
+	public async acquireChannel(func: (channel: Channel) => Promise<void>): Promise<void> {
+		let freeChannels = this.getFreeChannels();
+		if (freeChannels.length == 0) {
+			throw new ChannelBusyError();
+		}
+		const randomIndex = Math.floor(Math.random() * freeChannels.length);
+		const freeChannel = freeChannels[randomIndex];
+		freeChannel.state = 'busy';
+		try {
+			await func(freeChannel);
+		} finally {
+			this.releaseChannel(freeChannel);
+		}
+	}
+
+	public get status(): ChannelsPoolStatus {
+		const freeChannels = this.getFreeChannels().length;
+		return {
+			totalChannels: this.channels.length,
+			freeChannels: freeChannels,
+			busyChannels: this.channels.length - freeChannels,
+			//clone the mutable state to prevent accidentally external changes
+			channels: this.channels.map(channel => ({keyPair: channel.keyPair, state: channel.state}))
+		}
+	}
+
+	private releaseChannel(channel: Channel): void {
+		channel.state = 'free';
+	}
+}
+
+type ChannelState = 'busy' | 'free';
```

Comment:
yeah, I saw, I wasn't sure why we need it, what does it really means?
I mean in case of low balance we should fund it while submitting a tx, and I don't mind re-doing it, no? (in case of an error)
Actually, I think now we might have a problem in python, cause it release the channel only if it's not underfunded, so if top-up has failed, it will stuck in underfunded state, isn't it?

Reply:
Exactly, this is why the underfunded state exists. To keep the channel inactive if top-up fails

Reply:
but then we'll never retry to top-up, meaning single time error in top-up will pull it out from the queue